Feat. .Desktop Actions

harshaaaaaaaaaa opened the pull request:
Added .desktop actions popover.
Works with :
- ctrl+D to hide,
- alt+D to show,
- alt+D & alt+S to navigate in popover. & few more implementations
-  ctrl + S for favourite toggle


https://github.com/user-attachments/assets/d93ead93-1df6-48e7-90aa-8586c18460fb




Added libNotify for notification (Added to Favourite)

TO DO
- Add to favourite symbol
- App list sort improvement
- open in next workspace (works but couldn't open new workspace)

SoumyaRanjanPatnaik commented:
Can you resolve the conflicts?

SoumyaRanjanPatnaik commented on src/DialogWindow.vala:
```diff
@@ -131,10 +144,18 @@ namespace Ilia {
                         return true;
                     }
                     if (key.keyval == 'j' || key.keyval == 'n') { // Down - vim/emacs style
+                        if (is_on_help_page && help_scrolled_window != null) {
+                            scroll_help_window(true);
+                            return true;
+                        }
```
This can be handled cleanly with a seperate module or widget for the `HelpPage`.

SoumyaRanjanPatnaik commented on src/Util.vala:
Is there a standard implementation of fuzzy search available with GTK? You might be able to get faster / more accurate results using that. You can also use `ripgrep` for super fast, super accurate fuzzy finding. 

Your implementation seems to work well in my testing so if you can't find a good alternative, it looks good to me.

harshaaaaaaaaaa replied:
As far as I know, GTK itself doesn't provide a standard implementation of fuzzy search out of the box. 
External tools like "ripgrep" would definitely be a better fit in scenarios involving very large datasets (thousands of items) or full file content searches, where performance becomes critical. But for smaller lists or UI-level filtering, I think this approach holds up well.

Let me know if you think it’s worth exploring other options further!

SoumyaRanjanPatnaik commented on src/clipboard/ClipboardPage.vala:
The Clipboard page didn't work for me in my testing. Also instead of using gsettings to store the clipboard values, use a standard clipboard manager.

SoumyaRanjanPatnaik reviewed (changes requested):
Preliminary review. Haven't looked at the `DesktopAppPage.vala` file yet. LMK if you have any queries. Will do a more thorough review later.

harshaaaaaaaaaa commented on src/DialogWindow.vala:
```diff
@@ -5,8 +5,8 @@ namespace Ilia {
     public const int KEY_CODE_SUPER = 65515;
-    public const int KEY_CODE_UP = 65364;
-    public const int KEY_CODE_DOWN = 65362;
+    public const int KEY_CODE_UP = 65362;
+    public const int KEY_CODE_DOWN = 65364;
```
Yeah!

harshaaaaaaaaaa commented on src/Application.vala:
Not really, but accidental perhaps IDE setting. maybe

harshaaaaaaaaaa committed:
fix(dialog): correct navigation key codes and improve functionality

- Fix incorrect UP/DOWN key code constants (65362/65364 swapped)
- Improve cursor positioning after paste operations
- Enhance keyboard navigation handling for dialog pages
- Add proper Alt+D key event delegation to active pages
- Improve dialog resizing for both X11 and Wayland
- Add get_wm_name() method to SessionController interface

harshaaaaaaaaaa committed:
feat(search): implement fuzzy search algorithm for app filtering

- Add fuzzy_match_score() function for string matching
- Add character sequence matching
- Include fuzzy_match() helper with configurable threshold
- Replace old compare_desktop_apps sorting logic
- Improve search accuracy for application names and keywords

harshaaaaaaaaaa committed:
feat(apps): add desktop actions popover with favorites system

- Implement desktop actions popover accessible via Alt+D
- Add favorites system with Ctrl+S toggle and star indicators
- Support .desktop file actions and generic actions
- Include workspace management for next workspace launches
- Integrate fuzzy search with favorites prioritization
- Add GSettings storage for favorite apps list
- Enhance app sorting to prioritize favorites over launch history

harshaaaaaaaaaa committed:
feat(dialog): implement dedicated help page module

- Create standalone HelpPage class implementing DialogPage interface
- Add comprehensive keybinding documentation with TreeView
- Remove inline help implementation from DialogWindow

harshaaaaaaaaaa committed:
refractor(DialogWindow)

SoumyaRanjanPatnaik commented on src/DialogWindow.vala:
```diff
@@ -316,17 +315,25 @@ namespace Ilia {
+                case "clipboard":
+                    dialog_pages[0] = new ClipboardPage ();
+                    dialog_pages[0].initialize.begin(settings, arg_map, entry, this, this.wm_name, this.is_wayland);
+                    break;
```
Remove clipboard page.

SoumyaRanjanPatnaik commented on src/DialogWindow.vala:
```diff
@@ -341,6 +348,8 @@ namespace Ilia {
             dialog_pages[5] = new TrackerPage();
             dialog_pages[5].initialize.begin(settings, arg_map, entry, this, this.wm_name, this.is_wayland);
+            dialog_pages[6] = new ClipboardPage ();
+            dialog_pages[6].initialize.begin(settings, arg_map, entry, this, this.wm_name, this.is_wayland);
```
Remove clipboard page

SoumyaRanjanPatnaik commented on src/DialogWindow.vala:
```diff
@@ -362,6 +371,9 @@ namespace Ilia {
                 case "tracker":
                     start_page = 5;
                     break;
+                case "clipboard":
+                    start_page = 6;
+                    break;
```
Remove clipboard page.

SoumyaRanjanPatnaik commented on src/DialogWindow.vala:
```diff
@@ -316,17 +315,25 @@ namespace Ilia {
          */
         private int create_all_pages(HashTable<string, string ?> arg_map, string focus_page, ref uint start_page) {
-            int page_count = 6;
+            int page_count = 7; // increased for clipboard page
```
Should be 6 without clipboard page.

SoumyaRanjanPatnaik commented on src/DialogWindow.vala:
```diff
@@ -429,22 +401,46 @@ namespace Ilia {
+                return;
+            }
+            
+            if (dialog_pages[page_num] is HelpPage) { // On help page
```
HelpPage doesn't exist at this point.

SoumyaRanjanPatnaik commented on src/Util.vala:
```diff
@@ -95,63 +95,85 @@ namespace Ilia {
+            int pos = s_lower.index_of(p_lower);
+            if (pos == 0) {
+                // Starting match is best
+                return 95 + int.min(5, 5 * pattern.length / source.length);
+            } else {
+                // Penalize matches later in the string
+                return 80 + int.min(15, 20 * pattern.length / source.length) - int.min(15, pos / 2);
```
How did we arrive at these constants? are they part of the standard fuzzy searching algorithm? If not, can we extract them out to make it easier to understand their function as well as to customize?

SoumyaRanjanPatnaik replied:
Also seems like the penalty is added towards the end and will evaluate to zero if `pos == 0`, so maybe we can find a way to get rid of the conditional for a more consistent algorithm?

SoumyaRanjanPatnaik replied:
Maybe something like this makes more sense.
```vala
return SUBSTRING_BASE_SCORE+ LENGTH_COEFF * (pattern.length / source.length) - MAX_PENALTY * (pos / source.legth)
```

with the costants being defined as:
```vala
// Base score if the pattern is contained within the source string
int SUBSTRING_BASE_SCORE = 95;
// Coefficient of length for rewarding matches that are closest to the match
int LENGTH_COEFF = 5;
// Maximum penalty applicable matches later in the string
int MAX_PENALTY = 20;
```

Why this is better:
1. Named constants like I suggested are better for readability
2. The same formula can be used for both cases
3. When pattern is a substring, the ratio of `pattern_length/source_length` will never exceed one, so we can get rid of the `int.min` part.
4. The penalty section of the formula takes into account the length of the string as well, so we're not making any assumptions on till which point in the string we want to penalize.

SoumyaRanjanPatnaik commented on src/Util.vala:
```diff
@@ -95,63 +95,85 @@ namespace Ilia {
+        
+        // Calculate final score (0-100)
+        int score = (int)(match_ratio * 60.0 + length_ratio * 25.0 + consecutive_bonus);
+        return int.min(75, score); // Cap at 75 to rank below exact substring matches
```
This should be `int.max` if we want to cap the score.

Also it might be a better idea to scale the score by multiplying it instead of capping, since all the higher score matches will have the same score.

```vala
return score * (SUBSTRING_BASE_SCORE - MAX_PENALTY) / 100;
```

This would evaluate to 75 as per the constants in my previous comment.

SoumyaRanjanPatnaik commented on data/org.regolith-linux.ilia.gschema.xml:
```diff
@@ -8,6 +8,20 @@
+      <description>
+        clipboard-history persistance between sessions.
+      </description>
+    </key>
```
Remove clipboard related code.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -2,12 +2,12 @@ using Gtk;
 
 namespace Ilia {
     class DesktopAppPage : DialogPage, GLib.Object {
-        private static GLib.Timer app_timer;
```
Any reason for removing the app_timer?

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -400,5 +563,470 @@ namespace Ilia {
+                        }
+                        break;
+                        
+                    case "generic-folder":
+                        try {
+                            // Open the folder containing the desktop file
+                            string folder_path = Path.get_dirname(desktop_file_path);
```
I don't think most end users want to navigate to the desktop file. You can remove this case as well.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -400,5 +563,470 @@ namespace Ilia {
+                        }
+                        break;
+                        
+                    case "generic-new-workspace":
```
Extract the logic for this case into a seperate function for better readability

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -400,5 +563,470 @@ namespace Ilia {
+                            if (wm_cmd_newws != null) {
+                                // For i3/sway: switch to next workspace and launch app
+                                string launch_cmd = wm_cmd_newws + " workspace next && " + clean_exec;
+                                args = {"sh", "-c", launch_cmd};
+                            } else {
+                                args = {"sh", "-c", clean_exec};
```
Use `systemd-run` to launch applications. You can refer to `get_runner_app_info()` function on how to do that. See if you can use it directly to get the execute command.

SoumyaRanjanPatnaik replied:
Without `systemd-run` low memory situations can cause the entire session to crash.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -400,5 +563,470 @@ namespace Ilia {
+            try {
+                // Create empty AppLaunchContext instead of passing null
+                AppLaunchContext context = new AppLaunchContext();
+                app_info.launch_action(action_id, context);
```
Check if it is possible to make this use `systemd-run` as well. Anything that ilia tries to execute should ideally use that command.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -400,5 +563,470 @@ namespace Ilia {
+                        }
+                        break;
+                        
+                    case "generic-shortcut":
+                        try {
+                            // Create desktop shortcut
+                            string home_dir = Environment.get_home_dir();
```
This case is not relevant to i3 and sway since they don't have a desktop that can show icons

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -400,5 +563,470 @@ namespace Ilia {
+                        hide_app_actions();
+                        break;
+                
+                    case "generic-terminal":
+                        try {
+                            // Get the exec line from desktop file
+                            var key_file = new KeyFile();
```
Move the logic for this into a seperate function for better readability. Also make sure you use `systemd-run` to launch the terminal instance.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -245,14 +376,37 @@ namespace Ilia {
 
-                if (strval != null && strval.down().contains(query_string)) return true;
+                // Check if this is a favorite app (will be used for sorting)
+                model.get_value(iter, ITEM_VIEW_COLUMN_IS_FAVORITE, out app_info);
+                bool is_favorite = app_info.get_boolean();
```
I don't think this is necessary for filtering.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -245,14 +376,37 @@ namespace Ilia {
+                model.get_value(iter, ITEM_VIEW_COLUMN_IS_FAVORITE, out app_info);
+                bool is_favorite = app_info.get_boolean();
+
+                // Exact substring match - highest priority
+                if (strval != null && strval.down().contains(query_string)) {
+                    return true;
+                }
+
```
This case is handled by the fuzzy algorithm as well.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -245,14 +376,37 @@ namespace Ilia {
 
-                return strval != null && strval.down().contains(query_string);
+                // Try exact substring match on keywords
+                if (strval != null && strval.down().contains(query_string)) {
+                    return true;
+                }
```
This is covered by the fuzzy algorithm

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -245,14 +376,37 @@ namespace Ilia {
+
+                return false;
             } else {
-                return true;
+                return true; // Show all items when no query
```
nit: Move this to the top of the function with the guard condition:
```vala
if (strval == null) return true
```

Then you can remove the `strval != null`  condition from all the branches to make it more readable.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -192,27 +321,29 @@ namespace Ilia {
-            var app_a_id = app_a.get_id();
-            var app_b_id = app_b.get_id();
-
-            return compare_desktop_apps(
-                app_a_name,
-                app_b_name,
-                app_a_id,
-                app_b_id,
-                query_string,
-                launch_counts
-            );
```
I guess we could still have used this. favorites should be at the top, then apps sorted by launch counts and then alphabetical sort.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -16,6 +16,9 @@ namespace Ilia {
         // Roots of app dirs env var.  See https://github.com/flatpak/flatpak/issues/1286#issuecomment-354554684
         private const string XDG_DATA_DIRS = "XDG_DATA_DIRS";
 
+        // Backup suffix for desktop files
+        private const string BACKUP_SUFFIX = ".ilia-backup";
+
```
I guess this is not required, since we don't want to edit desktop files.

SoumyaRanjanPatnaik commented on test/UtilTest.vala:
```diff
@@ -55,11 +56,20 @@ public void test_compare_desktop_apps() {
 
     Test.add_func ("/test_systemd_escape", test_systemd_escape);
-    Test.add_func ("/test_compare_desktop_apps", test_compare_desktop_apps);
+//    Test.add_func ("/test_compare_desktop_apps", test_compare_desktop_apps);
```
Don't comment out code. The commit history will contain this.

SoumyaRanjanPatnaik commented on src/DialogWindow.vala:
I think this commit fixes some of my comments regarding the clipboard page. You can squash this commit with the commit where I provided the review for removing this code.

Also the commit name `refractor(DialogWindow)` neither describes the code change, nor does it follow the conventional commit format.

SoumyaRanjanPatnaik commented on src/Util.vala:
```diff
@@ -107,26 +107,39 @@ namespace Ilia {
             } else {
-                // Penalize matches later in the string
-                return 80 + int.min(15, 20 * pattern.length / source.length) - int.min(15, pos / 2);
+                if (pos > 0 && (s_lower[pos - 1] == ' ' || s_lower[pos - 1] == '-' || s_lower[pos - 1] == '_')) {
+                    return 85 + int.min(10, 10 * pattern.length / source.length);
+                }
+                // Regular substring match - penalize based on position
+                return 75 + int.min(10, 15 * pattern.length / source.length) - int.min(10, pos / 3);
```
Refer to my comment on the old commit. See if you can generalize these cases and make the formula more readable. Also is there a specific reason for prioritizing cases with _,- or space?

SoumyaRanjanPatnaik commented on src/Util.vala:
```diff
@@ -107,26 +107,39 @@ namespace Ilia {
+        // Check for acronym match (first letters of words)
+        if (check_acronym_match(s_lower, p_lower)) {
+            return 70 + int.min(15, 20 * pattern.length / source.length);
+        }
```
You can skip this check also. We want to minimize the number of checks we do.

SoumyaRanjanPatnaik commented on src/Util.vala:
```diff
@@ -137,6 +150,16 @@ namespace Ilia {
+                    if (j > 0 && (s_lower[j - 1] == ' ' || s_lower[j - 1] == '-' || s_lower[j - 1] == '_')) {
+                        word_boundary_bonus += 5;
+                    }
+                    
```
Again not sure how this helps.

SoumyaRanjanPatnaik commented on src/apps/DesktopAppPage.vala:
```diff
@@ -332,20 +340,70 @@ namespace Ilia {
             }
 
-            // If both are favorites or both are not favorites, sort alphabetically first
+            // If scores are equal or not searching, sort alphabetically
```
You can move this code into the function `compare_desktop_apps` that was previously used here.

harshaaaaaaaaaa committed:
fix(gschema): remove clipboard history key from schema
